lib/HLSL/DxilAddPixelHitInstrumentation.cpp

```diff
@@ -116,10 +121,50 @@ bool DxilAddPixelHitInstrumentation::runOnModule(Module &M)
   auto EntryPointFunction = DM.GetEntryFunction();
 
   auto & EntryBlock = EntryPointFunction->getEntryBlock();
-  bool HaveInsertedUAV = false;
 
   CallInst *HandleForUAV;
-
+  {
+    IRBuilder<> Builder(DM.GetEntryFunction()->getEntryBlock().getFirstInsertionPt());
+    
+    unsigned int UAVResourceHandle = static_cast<unsigned int>(DM.GetUAVs().size());
+
+    // Set up a UAV with structure of a single int
+    SmallVector<llvm::Type*, 1> Elements{ Type::getInt32Ty(Ctx) };
+    llvm::StructType *UAVStructTy = llvm::StructType::create(Elements, "class.RWStructuredBuffer");
+    std::unique_ptr<DxilResource> pUAV = llvm::make_unique<DxilResource>();
+    pUAV->SetGlobalName("PIX_CountUAVName");
+    pUAV->SetGlobalSymbol(UndefValue::get(UAVStructTy->getPointerTo()));
+    pUAV->SetID(UAVResourceHandle);
+    pUAV->SetSpaceID((unsigned int)-2); // This is the reserved-for-tools register space
+    pUAV->SetSampleCount(1);
+    pUAV->SetGloballyCoherent(false);
+    pUAV->SetHasCounter(false);
+    pUAV->SetCompType(CompType::getI32());
+    pUAV->SetLowerBound(0);
+    pUAV->SetRangeSize(1);
+    pUAV->SetKind(DXIL::ResourceKind::RawBuffer);
+
+    auto pAnnotation = DM.GetTypeSystem().AddStructAnnotation(UAVStructTy);
+    pAnnotation->GetFieldAnnotation(0).SetCBufferOffset(0);
+    pAnnotation->GetFieldAnnotation(0).SetCompType(hlsl::DXIL::ComponentType::I32);
+    pAnnotation->GetFieldAnnotation(0).SetFieldName("count");
+
+    ID = DM.AddUAV(std::move(pUAV));
+
+    assert(ID == UAVResourceHandle);
+
+    // Create handle for the newly-added UAV
+    Function* CreateHandleOpFunc = HlslOP->GetOpFunc(DXIL::OpCode::CreateHandle, Type::getVoidTy(Ctx));
+    Constant* CreateHandleOpcodeArg = HlslOP->GetU32Const((unsigned)DXIL::OpCode::CreateHandle);
+    Constant* UAVVArg = HlslOP->GetI8Const(static_cast<std::underlying_type<DxilResourceBase::Class>::type>(DXIL::ResourceClass::UAV));
```

Comment:
Extra 'V' in UAVVArg?
Also, what's wrong with a simple cast (`(unsigned)DXIL::ResourceClass::UAV` or `uint8_t`) here, such as used for DXIL::OpCode::CreateHandle?

Reply:
Thanks!